### PR TITLE
bardi_0.4.0_egress

### PR DESCRIPTION
**Pull request by dakotamurdock**
Merge Egress files

Updates:

- Option to retain input fields in each step that alters the data
- Changes PostProcessor to VocabEncoder
- Introduces tokenizers

### CODE REVIEW

File: bardi/data/data_handlers.py [hunk 1]
```diff
@@ -126,9 +126,7 @@ def from_file(
         if min_batches:
             # If a min_batches is specified, data from the file(s) are split
             # into a list of smaller PyArrow Tables
-            tables = chunk_pyarrow_table(
-                data=table, row_count=row_count, min_batches=min_batches
-            )
+            tables = chunk_pyarrow_table(data=table, row_count=row_count, min_batches=min_batches)
```
Comment (dakotamurdock) [inline bardi/data/data_handlers.py:129]: formatting

File: bardi/data/data_handlers.py [hunk 2]
```diff
@@ -183,9 +181,7 @@ def from_duckdb(path: str, query: str, min_batches: int = None) -> Dataset:
     if min_batches:
         # If a min_batches is specified, data from the file(s) are split into a
         # list of smaller PyArrow Tables
-        tables = chunk_pyarrow_table(
-            data=table, row_count=row_count, min_batches=min_batches
-        )
+        tables = chunk_pyarrow_table(data=table, row_count=row_count, min_batches=min_batches)
```
Comment (dakotamurdock) [inline bardi/data/data_handlers.py:184]: formatting

File: bardi/data/data_handlers.py [hunk 3]
```diff
@@ -239,9 +235,7 @@ def from_pandas(df: pd.DataFrame, min_batches: int = None) -> Dataset:
     if min_batches:
         # If a min_batches is specified, data from the file(s) are split into
         # a list of smaller PyArrow Tables
-        tables = chunk_pyarrow_table(
-            data=table, row_count=row_count, min_batches=min_batches
-        )
+        tables = chunk_pyarrow_table(data=table, row_count=row_count, min_batches=min_batches)
```
Comment (dakotamurdock) [inline bardi/data/data_handlers.py:238]: formatting

File: bardi/data/data_handlers.py [hunk 4]
```diff
@@ -288,9 +282,7 @@ def from_pyarrow(table: pa.Table, min_batches: int = None) -> Dataset:
     if min_batches:
         # If a min_batches is specified, data from the file(s) are split into a
         # list of smaller PyArrow Tables
-        tables = chunk_pyarrow_table(
-            data=table, row_count=row_count, min_batches=min_batches
-        )
+        tables = chunk_pyarrow_table(data=table, row_count=row_count, min_batches=min_batches)
```
Comment (dakotamurdock) [inline bardi/data/data_handlers.py:285]: formatting

File: bardi/data/data_handlers.py [hunk 5]
```diff
@@ -373,13 +365,13 @@ def to_pandas(table: pa.Table) -> pd.DataFrame:
...
     """Return data as a polars DataFrame
-    
+
     Parameters
     ----------
 
     table : PyArrow.Table
         Table of data you want to convert to a Pandas DataFrame
-    
+
```
Comment (dakotamurdock) [inline bardi/data/data_handlers.py:374]: formatting

File: bardi/data/data_handlers.py [hunk 6]
```diff
@@ -430,11 +422,11 @@ def write_file(data: pa.Table, path: str, format: str, *args, **kwargs) -> None:
             schema = csv_df.schema
             csv_df = csv_df.with_columns(
                 [
-                    pl.format(
-                        "[{}]", pl.col(field).cast(pl.List(pl.Utf8)).list.join(", ")
-                    ).alias(field)
+                    pl.format("[{}]", pl.col(field).cast(pl.List(pl.Utf8)).list.join(", ")).alias(
+                        field
+                    )
```
Comment (dakotamurdock) [inline bardi/data/data_handlers.py:427]: formatting

File: bardi/nlp_engineering/label_processor.py
Comment (dakotamurdock) [inline bardi/nlp_engineering/label_processor.py:1]: formatting and add retain inputs

File: bardi/nlp_engineering/normalizer.py
Comment (dakotamurdock) [inline bardi/nlp_engineering/normalizer.py:1]: formatting and add retain inputs

File: bardi/nlp_engineering/pre_tokenizer.py
Comment (dakotamurdock) [inline bardi/nlp_engineering/pre_tokenizer.py:1]: formatting and add retain inputs

File: bardi/nlp_engineering/splitter.py [hunk 7]
```diff
@@ -299,17 +291,13 @@ def run(self, data: pa.Table, artifacts: dict = None) -> Tuple[pa.Table, dict]:
...
-                        .alias("composite_record_id")
-                    ]
+                    [pl.concat_str([*self.unique_record_cols]).hash().alias("composite_record_id")]
                 )
                 .with_columns(
                     [
                         pl.col("composite_record_id")
                         .cast(pl.Utf8())
-                        .map_dict(self.split_mapping)
+                        .replace(self.split_mapping, default=self.default_split_value)
```
Comment (dakotamurdock) [inline bardi/nlp_engineering/splitter.py:300]: change in polars api

File: bardi/nlp_engineering/splitter.py [hunk 8]
```diff
@@ -354,7 +344,7 @@ def run(self, data: pa.Table, artifacts: dict = None) -> Tuple[pa.Table, dict]:
             # mapped to assigned value back to the name of
             # the split
             splits = distinct_groups.with_columns(
-                split_labels.map_dict(split_name_maps).alias("split")
+                split_labels.replace(split_name_maps).alias("split")
```
Comment (dakotamurdock) [inline bardi/nlp_engineering/splitter.py:347]: change in polars api

File: bardi/nlp_engineering/splitter.py [hunk 9]
```diff
@@ -323,16 +311,18 @@ def run(self, data: pa.Table, artifacts: dict = None) -> Tuple[pa.Table, dict]:
...
+            # Generate a permutation of indices of distinct groups.
             permuted_indices = np.random.permutation(group_count)
 
             # The splits assignment has to be done with a number
             # so a mapping must be created from a number to the
             # desired split name
-            split_name_maps = {
-                i: key for i, key in enumerate(self.split_proportions.keys())
-            }
+            split_name_maps = {i: key for i, key in enumerate(self.split_proportions.keys())}
```
Comment (dakotamurdock) [inline bardi/nlp_engineering/splitter.py:325]: formatting

File: bardi/nlp_engineering/regex_library/regex_set.py
Comment (dakotamurdock) [inline bardi/nlp_engineering/regex_library/regex_set.py:1]: formatting, introduction of lowercase and no substitution configuration for regex substitutions

File: bardi/nlp_engineering/splitter.py
Comment (dakotamurdock) [inline bardi/nlp_engineering/splitter.py:1]: formatting, change map_dict to replace per change in polars api

File: bardi/nlp_engineering/vocab_encoder.py
Comment (dakotamurdock) [inline bardi/nlp_engineering/vocab_encoder.py:1]: rename post processor to vocab encoder, add retain inputs

File: bardi/pipeline.py
Comment (dakotamurdock) [inline bardi/pipeline.py:1]: formatting and docstrings

File: docs/source/advanced.rst
Comment (dakotamurdock) [inline docs/source/advanced.rst:1]: simplified imports in code examples

**Comment by mslovett21** [comment]
Looks good.

**Review by mslovett21 — APPROVED** [review]
Looks good.

**Pull request merged**